src/backend/cuda/kernel/homography.hpp

```diff
@@ -0,0 +1,700 @@
+/*******************************************************
+ * Copyright (c) 2015, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <af/defines.h>
+#include <dispatch.hpp>
+#include <err_cuda.hpp>
+#include <debug_cuda.hpp>
+#include <memory.hpp>
+#include "ireduce.hpp"
+#include "reduce.hpp"
+#include "sort.hpp"
+
+#include <cfloat>
+
+#include <iostream>
+
+namespace cuda
+{
+
+namespace kernel
+{
+
+template<typename T>
+__device__ T sq(T a)
+{
+    return a * a;
+}
+
+template<typename T>
+struct EPS
+{
+    __device__ T eps() { return FLT_EPSILON; }
+};
+
+template<>
+struct EPS<float>
+{
+    __device__ static float eps() { return FLT_EPSILON; }
+};
+
+template<>
+struct EPS<double>
+{
+    __device__ static double eps() { return DBL_EPSILON; }
+};
+
+#define RANSACConfidence 0.99f
+#define LMEDSConfidence 0.99f
+#define LMEDSOutlierRatio 0.4f
+
+
+template<typename T>
+__device__ void JacobiSVD(T* S, T* V, int m, int n)
+{
+    const int iterations = 30;
+
+    int tid_x = threadIdx.x;
+    int bsz_x = blockDim.x;
+    int tid_y = threadIdx.y;
+    int gid_y = blockIdx.y * blockDim.y + tid_y;
+
+    __shared__ T acc[512];
+    T* acc1 = acc;
+    T* acc2 = acc + 256;
+
+    __shared__ T s_S[16*81];
+    __shared__ T s_V[16*81];
+    __shared__ T d[16*9];
+
+    for (int i = 0; i <= 4; i++)
+        s_S[tid_y * 81 + i*bsz_x + tid_x] = S[gid_y * 81 + i*bsz_x + tid_x];
+    if (tid_x == 0)
+        s_S[tid_y * 81 + 80] = S[gid_y * 81 + 80];
+    __syncthreads();
+
+    // Copy first 80 elements
+    for (int i = 0; i <= 4; i++) {
+        T t = s_S[tid_y*81 + tid_x+i*bsz_x];
+        acc1[tid_y*bsz_x + tid_x] += t*t;
+    }
+    if (tid_x < 8)
+        acc1[tid_y*16 + tid_x] += acc1[tid_y*16 + tid_x+8];
+    __syncthreads();
```

Comment:
They may not be necessary here but it is safer and follows closely the OpenCL code, plus, this won't make too much of a difference in performance.